"""
Application GUI pour la r√©solution de labyrinthes avec A* et Dijkstra.
"""
import sys
import time
import numpy as np
from PySide6.QtWidgets import (QApplication, QMainWindow, QGraphicsScene, 
                               QGraphicsRectItem, QMessageBox)
from PySide6.QtCore import Qt, QRectF
from PySide6.QtGui import QBrush, QColor, QPen

from ui_maze_window import Ui_MainWindow
from Maze import Maze
from main import create_complete_maze


class MazeWindow(QMainWindow):
    """Fen√™tre principale de l'application."""
    
    def __init__(self):
        super().__init__()
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)
        
        # Variables pour le labyrinthe
        self.maze = None
        self.scene = QGraphicsScene()
        self.ui.mazeView.setScene(self.scene)
        self.cell_size = 30
        self.path_astar = None
        self.path_dijkstra = None
        
        # Couleurs
        self.color_free = QColor(255, 255, 255)
        self.color_obstacle = QColor(50, 50, 50)
        self.color_start = QColor(0, 255, 0)
        self.color_goal = QColor(255, 0, 0)
        self.color_path_astar = QColor(100, 150, 255)
        self.color_path_dijkstra = QColor(255, 150, 100)
        self.color_bonus = QColor(255, 255, 0)
        
        # Connecter les signaux
        self.ui.generateButton.clicked.connect(self.generate_maze)
        self.ui.solveAStarButton.clicked.connect(self.solve_astar)
        self.ui.solveDijkstraButton.clicked.connect(self.solve_dijkstra)
        self.ui.compareButton.clicked.connect(self.compare_algorithms)
        self.ui.clearButton.clicked.connect(self.clear_path)
        self.ui.actionQuitter.triggered.connect(self.close)
        self.ui.actionAPropos.triggered.connect(self.show_about)
        
        # G√©n√©rer un labyrinthe par d√©faut
        self.generate_maze()
    
    def generate_maze(self):
        """G√©n√®re un nouveau labyrinthe selon les param√®tres."""
        width = self.ui.widthSpinBox.value()
        height = self.ui.heightSpinBox.value()
        
        # D√©terminer le type d'obstacles
        obstacle_type_index = self.ui.obstacleTypeCombo.currentIndex()
        obstacle_types = ["random", "vertical_walls", "horizontal_walls", 
                         "maze_pattern", "none"]
        obstacle_type = obstacle_types[obstacle_type_index]
        
        density = self.ui.densitySpinBox.value()
        step_cost = self.ui.stepCostSpinBox.value()
        goal_reward = self.ui.goalRewardSpinBox.value()
        add_bonuses = self.ui.bonusCheckBox.isChecked()
        num_bonuses = self.ui.numBonusSpinBox.value()
        bonus_value = self.ui.bonusValueSpinBox.value()
        
        # Cr√©er le labyrinthe
        if obstacle_type == "none":
            self.maze = Maze(width, height, start=(0, 0), 
                           goal=(height-1, width-1))
            # Initialiser les r√©compenses
            self.maze.rewards = np.full((height, width), step_cost, dtype=float)
            self.maze.set_reward(*self.maze.goal, goal_reward)
        else:
            self.maze = create_complete_maze(
                width, height,
                obstacle_type=obstacle_type,
                obstacle_density=density,
                step_cost=step_cost,
                goal_reward=goal_reward,
                add_bonuses=add_bonuses,
                num_bonuses=num_bonuses,
                bonus_value=bonus_value
            )
        
        # R√©initialiser les chemins
        self.path_astar = None
        self.path_dijkstra = None
        
        # Afficher le labyrinthe
        self.draw_maze()
        
        # Afficher les infos
        self.ui.statsText.clear()
        self.ui.statsText.append(f"Labyrinthe g√©n√©r√©: {width}x{height}")
        self.ui.statsText.append(f"D√©part: {self.maze.start}")
        self.ui.statsText.append(f"Arriv√©e: {self.maze.goal}")
    
    def draw_maze(self):
        """Dessine le labyrinthe dans la vue graphique."""
        if self.maze is None:
            return
        
        self.scene.clear()
        
        # Calculer la taille des cellules
        view_width = self.ui.mazeView.width() - 10
        view_height = self.ui.mazeView.height() - 10
        self.cell_size = min(view_width // self.maze.width, 
                            view_height // self.maze.height, 40)
        
        # Dessiner chaque cellule
        for i in range(self.maze.height):
            for j in range(self.maze.width):
                x = j * self.cell_size
                y = i * self.cell_size
                
                # D√©terminer la couleur
                if (i, j) == self.maze.start:
                    color = self.color_start
                elif (i, j) == self.maze.goal:
                    color = self.color_goal
                elif self.maze.grid[i, j] == 1:
                    color = self.color_obstacle
                elif self.maze.rewards[i, j] > 0:
                    color = self.color_bonus
                else:
                    color = self.color_free
                
                # Dessiner le rectangle
                rect = QGraphicsRectItem(QRectF(x, y, self.cell_size, self.cell_size))
                rect.setBrush(QBrush(color))
                rect.setPen(QPen(Qt.black, 1))
                self.scene.add_item(rect)
        
        # Dessiner les chemins s'ils existent
        if self.path_astar is not None:
            self.draw_path(self.path_astar, self.color_path_astar, 0.3)
        
        if self.path_dijkstra is not None:
            self.draw_path(self.path_dijkstra, self.color_path_dijkstra, 0.3)
    
    def draw_path(self, path, color, alpha):
        """Dessine un chemin sur le labyrinthe."""
        if path is None or len(path) == 0:
            return
        
        color_with_alpha = QColor(color)
        color_with_alpha.setAlphaF(alpha)
        
        for (i, j) in path:
            # Ne pas redessiner le d√©part et l'arriv√©e
            if (i, j) == self.maze.start or (i, j) == self.maze.goal:
                continue
            
            x = j * self.cell_size
            y = i * self.cell_size
            
            rect = QGraphicsRectItem(QRectF(x + 2, y + 2, 
                                           self.cell_size - 4, 
                                           self.cell_size - 4))
            rect.setBrush(QBrush(color_with_alpha))
            rect.setPen(QPen(color, 2))
            self.scene.addItem(rect)
    
    def solve_astar(self):
        """R√©sout le labyrinthe avec A*."""
        if self.maze is None:
            return
        
        start_time = time.time()
        self.path_astar = self.maze.solve()
        elapsed_time = time.time() - start_time
        
        if self.path_astar is None:
            QMessageBox.warning(self, "Aucun chemin", 
                              "Aucun chemin n'a √©t√© trouv√© avec A* !")
            self.ui.statsText.append("\n‚ùå A* : Aucun chemin trouv√©")
        else:
            # Calculer le co√ªt du chemin
            cost = self.calculate_path_cost(self.path_astar)
            
            self.ui.statsText.append(f"\n‚úÖ A* : Chemin trouv√©")
            self.ui.statsText.append(f"   Longueur: {len(self.path_astar)} cellules")
            self.ui.statsText.append(f"   Co√ªt total: {cost:.2f}")
            self.ui.statsText.append(f"   Temps: {elapsed_time*1000:.2f} ms")
            
            self.draw_maze()
    
    def solve_dijkstra(self):
        """R√©sout le labyrinthe avec Dijkstra."""
        if self.maze is None:
            return
        
        start_time = time.time()
        self.path_dijkstra = self.maze.solve_dijkstra()
        elapsed_time = time.time() - start_time
        
        if self.path_dijkstra is None:
            QMessageBox.warning(self, "Aucun chemin", 
                              "Aucun chemin n'a √©t√© trouv√© avec Dijkstra !")
            self.ui.statsText.append("\n‚ùå Dijkstra : Aucun chemin trouv√©")
        else:
            # Calculer le co√ªt du chemin
            cost = self.calculate_path_cost(self.path_dijkstra)
            
            self.ui.statsText.append(f"\n‚úÖ Dijkstra : Chemin trouv√©")
            self.ui.statsText.append(f"   Longueur: {len(self.path_dijkstra)} cellules")
            self.ui.statsText.append(f"   Co√ªt total: {cost:.2f}")
            self.ui.statsText.append(f"   Temps: {elapsed_time*1000:.2f} ms")
            
            self.draw_maze()
    
    def compare_algorithms(self):
        """Compare les deux algorithmes."""
        if self.maze is None:
            return
        
        self.ui.statsText.clear()
        self.ui.statsText.append("=== Comparaison A* vs Dijkstra ===\n")
        
        # A*
        start_time = time.time()
        self.path_astar = self.maze.solve()
        time_astar = time.time() - start_time
        
        # Dijkstra
        start_time = time.time()
        self.path_dijkstra = self.maze.solve_dijkstra()
        time_dijkstra = time.time() - start_time
        
        # Afficher les r√©sultats
        if self.path_astar is None:
            self.ui.statsText.append("‚ùå A* : Aucun chemin trouv√©\n")
        else:
            cost_astar = self.calculate_path_cost(self.path_astar)
            self.ui.statsText.append(f"‚úÖ A* :")
            self.ui.statsText.append(f"   Longueur: {len(self.path_astar)} cellules")
            self.ui.statsText.append(f"   Co√ªt: {cost_astar:.2f}")
            self.ui.statsText.append(f"   Temps: {time_astar*1000:.2f} ms\n")
        
        if self.path_dijkstra is None:
            self.ui.statsText.append("‚ùå Dijkstra : Aucun chemin trouv√©\n")
        else:
            cost_dijkstra = self.calculate_path_cost(self.path_dijkstra)
            self.ui.statsText.append(f"‚úÖ Dijkstra :")
            self.ui.statsText.append(f"   Longueur: {len(self.path_dijkstra)} cellules")
            self.ui.statsText.append(f"   Co√ªt: {cost_dijkstra:.2f}")
            self.ui.statsText.append(f"   Temps: {time_dijkstra*1000:.2f} ms\n")
        
        # Analyse comparative
        if self.path_astar is not None and self.path_dijkstra is not None:
            self.ui.statsText.append("üìä Analyse comparative :")
            
            if len(self.path_astar) == len(self.path_dijkstra):
                self.ui.statsText.append("   M√™me longueur de chemin")
            else:
                faster = "A*" if len(self.path_astar) < len(self.path_dijkstra) else "Dijkstra"
                self.ui.statsText.append(f"   {faster} trouve un chemin plus court")
            
            if time_astar < time_dijkstra:
                speedup = time_dijkstra / time_astar
                self.ui.statsText.append(f"   A* est {speedup:.2f}x plus rapide")
            else:
                speedup = time_astar / time_dijkstra
                self.ui.statsText.append(f"   Dijkstra est {speedup:.2f}x plus rapide")
            
            self.ui.statsText.append(f"\nüí° Complexit√© th√©orique :")
            self.ui.statsText.append(f"   A* : O((V+E) log V) avec heuristique")
            self.ui.statsText.append(f"   Dijkstra : O((V+E) log V) sans heuristique")
            self.ui.statsText.append(f"\n   A* explore moins de noeuds gr√¢ce √† l'heuristique")
        
        self.draw_maze()
    
    def calculate_path_cost(self, path):
        """Calcule le co√ªt total d'un chemin."""
        if path is None or len(path) == 0:
            return 0
        
        cost = 0
        for (i, j) in path:
            cost -= self.maze.rewards[i, j]
        
        return cost
    
    def clear_path(self):
        """Efface les chemins affich√©s."""
        self.path_astar = None
        self.path_dijkstra = None
        self.draw_maze()
        self.ui.statsText.append("\nChemins effac√©s")
    
    def show_about(self):
        """Affiche la bo√Æte de dialogue √Ä propos."""
        QMessageBox.about(self, "√Ä propos",
                         "R√©solution de Labyrinthe\n\n"
                         "TP1 - Robotique : R√©soudre un labyrinthe avec A*\n\n"
                         "Impl√©mentation des algorithmes :\n"
                         "‚Ä¢ Dijkstra : Algorithme de plus court chemin\n"
                         "‚Ä¢ A* : Algorithme avec heuristique\n\n"
                         "Quang-Trung Luu\n"
                         "Universit√© Paris-Saclay")


def main():
    """Point d'entr√©e de l'application."""
    app = QApplication(sys.argv)
    window = MazeWindow()
    window.show()
    sys.exit(app.exec())


if __name__ == "__main__":
    main()
